File: server/src/controllers/message.controller.js
import mongoose from "mongoose";
import Message from "../models/message.model.js";
import Conversation from "../models/conversation.model.js";
import User from "../models/user.model.js";
import admin from "../config/firebase.js";
import cloudinary from "../config/cloudinary.js"; 
import * as socketModule from "../socket/index.js"; 
import { generateAIResponse } from "../services/ai.service.js"; 
import apiResponse from "../utils/apiResponse.js";
import apiError from "../utils/apiError.js";
import asyncHandler from "../utils/asyncHandler.js";
import { MESSAGE_EDIT_WINDOW_MS } from "../constants/constants.js";

// Send silent push notification using FCM
const sendSilentPush = async (token, senderName, ciphertext, conversationId) => {
    if (!token) return;
    try {
        const payload = {
            data: {
                senderName,
                ciphertext: ciphertext || "",
                conversationId: conversationId || ""
            },
            token
        };
        await admin.messaging().send(payload);
    } catch (pushError) {
        console.error("Failed to send silent push notification:", pushError);
    }
};

// Retrieve messages in a conversation — bounded to 50 most recent
export const getMessages = asyncHandler(async (req, res, next) => {
    const { conversationId } = req.params;
    const userId = req.user?._id;

    if (!userId || !mongoose.Types.ObjectId.isValid(userId)) {
        throw new apiError(401, "Unauthorized");
    }

    if (!conversationId || !mongoose.Types.ObjectId.isValid(conversationId)) {
        throw new apiError(400, "Invalid conversation ID format.");
    }

    const conversation = await Conversation.findById(conversationId).lean();
    if (!conversation) {
        throw new apiError(404, "Conversation not found");
    }

    const isParticipant = conversation.participants.some(p => p.toString() === userId.toString());
    if (!isParticipant) {
        throw new apiError(403, "You are not authorized to view messages in this conversation");
    }

    const filter = { conversationId };
    filter.deletedForMe = { $ne: userId };

    // Fetch the 50 most recent messages
    const messages = await Message.find(filter)
        .sort({ createdAt: -1 }) // Get newest first
        .limit(50)
        .lean();

    // Reverse the array so the frontend receives them in chronological order (oldest to newest)
    const chronologicalMessages = messages.reverse();

    res.status(200).json(new apiResponse(200, chronologicalMessages, "Messages fetched successfully"));
});

// Clear all message history for conversations the user is in
export const clearMessages = asyncHandler(async (req, res, next) => {
    const userId = req.user?._id;

    if (!userId || !mongoose.Types.ObjectId.isValid(userId)) {
        throw new apiError(401, "Unauthorized");
    }

    const conversations = await Conversation.find({ participants: userId })
        .select("_id")
        .lean();
    const conversationIds = conversations.map((c) => c._id);

    if (conversationIds.length === 0) {
        return res.status(200).json(new apiResponse(200, "No messages to clear.", { deletedCount: 0 }));
    }

    const result = await Message.deleteMany({ conversationId: { $in: conversationIds } });

    await Conversation.updateMany(
        { _id: { $in: conversationIds } },
        { $set: { lastMessageId: null } }
    );

    return res.status(200).json(
        new apiResponse(200, "Chat history cleared successfully.", {
            deletedCount: result.deletedCount || 0,
        })
    );
});

// Upload raw encrypted attachment to Cloudinary
export const uploadAttachment = asyncHandler(async (req, res, next) => {
    if (!req.file || !req.file.buffer) {
        throw new apiError(400, "No file provided.");
    }

    const result = await new Promise((resolve, reject) => {
        const uploadStream = cloudinary.uploader.upload_stream(
            { resource_type: "raw", folder: "zync_secure_media" },
            (error, uploaded) => (error ? reject(error) : resolve(uploaded))
        );
        uploadStream.end(req.file.buffer);
    });

    return res.status(200).json(
        new apiResponse(200, "File uploaded successfully", { url: result.secure_url })
    );
});

// Send a message and optional AI integration response
export const sendMessage = asyncHandler(async (req, res, next) => {
    const conversationId = req.params.conversationId || req.params.id || req.body.conversationId;
    const { text, image, attachmentUrl, attachmentType, attachmentMime } = req.body;
    const senderId = req.user?._id;

    if (!senderId || !mongoose.Types.ObjectId.isValid(senderId)) {
        throw new apiError(401, "Unauthorized");
    }

    if (!conversationId || !mongoose.Types.ObjectId.isValid(conversationId)) {
        throw new apiError(400, "Conversation ID is missing or invalid.");
    }

    const hasText = typeof text === "string" && text.trim();
    if (!hasText && !image && !attachmentUrl) {
        throw new apiError(400, "Message must contain text, an image, or an attachment.");
    }

    const conversation = await Conversation.findById(conversationId);
    if (!conversation) {
        throw new apiError(404, "Conversation not found");
    }

    const isParticipant = conversation.participants.some(p => p.toString() === senderId.toString());
    if (!isParticipant) {
        throw new apiError(403, "You are not authorized to send messages");
    }

    let imageUrl = "";
    if (image) {
        const uploadResponse = await cloudinary.uploader.upload(image, {
            folder: "zync_messages",
        });
        imageUrl = uploadResponse.secure_url;
    }

    const newMessage = await Message.create({
        conversationId,
        senderId,
        text: text ? text.trim() : "",
        imageUrl,
        attachmentUrl: attachmentUrl || "",
        attachmentType: attachmentType || "",
        attachmentMime: attachmentMime || ""
    });

    conversation.lastMessageAt = new Date();
    conversation.lastMessageId = newMessage._id;
    await conversation.save();

    const otherParticipantIds = conversation.participants.filter(
        (pId) => pId.toString() !== senderId.toString()
    );

    const receivers = await User.find({ _id: { $in: otherParticipantIds } }).lean();
    const io = socketModule.io || (socketModule.getIO && socketModule.getIO());

    for (const receiver of receivers) {
        if (receiver.isAI) {
            let plainTextPrompt = text || "";
            if (image) plainTextPrompt += "\n[User also attached an image payload]";

            const aiResponseText = await generateAIResponse(plainTextPrompt);

            const aiMessage = await Message.create({
                senderId: receiver._id,
                conversationId: conversation._id,
                text: aiResponseText,
                imageUrl: ""
            });

            conversation.lastMessageAt = new Date();
            conversation.lastMessageId = aiMessage._id;
            await conversation.save();

            if (io) {
                const aiPayload = {
                    ...aiMessage.toObject(),
                    conversationId: aiMessage.conversationId.toString(),
                    senderId: aiMessage.senderId.toString(),
                };
                io.to(senderId.toString()).emit("newMessage", aiPayload);
            }

            if (req.user && req.user.fcmToken) {
                await sendSilentPush(
                    req.user.fcmToken,
                    receiver.displayName || receiver.username || "AI Assistant",
                    aiMessage.text,
                    conversation._id.toString()
                );
            }
        } else {
            if (io) {
                const payload = {
                    ...newMessage.toObject(),
                    conversationId: newMessage.conversationId.toString(),
                    senderId: newMessage.senderId.toString(),
                };
                io.to(receiver._id.toString()).emit("newMessage", payload);
            }

            if (receiver.fcmToken) {
                await sendSilentPush(
                    receiver.fcmToken,
                    req.user.displayName || req.user.username || "Someone",
                    newMessage.text,
                    conversation._id.toString()
                );
            }
        }
    }

    res.status(201).json(new apiResponse(201, "Message sent successfully", newMessage));
});

// Edit an existing message if within the 15-minute edit window
export const editMessage = asyncHandler(async (req, res, next) => {
    const { messageId } = req.params;
    const { text } = req.body;
    const userId = req.user?._id;

    if (!userId || !mongoose.Types.ObjectId.isValid(userId)) {
        throw new apiError(401, "Unauthorized");
    }

    if (!messageId || !mongoose.Types.ObjectId.isValid(messageId)) {
        throw new apiError(400, "Invalid message ID format.");
    }

    const message = await Message.findById(messageId);
    if (!message) {
        throw new apiError(404, "Message not found");
    }

    if (message.senderId.toString() !== userId.toString()) {
        throw new apiError(403, "You can only edit your own messages");
    }

    const timeDiff = Date.now() - new Date(message.createdAt).getTime();
    if (timeDiff >= MESSAGE_EDIT_WINDOW_MS) {
        throw new apiError(403, "Message cannot be edited after the edit window has elapsed");
    }

    message.text = text ? text.trim() : "";
    message.isEdited = true;
    await message.save();

    const conversation = await Conversation.findById(message.conversationId);
    if (conversation) {
        const io = socketModule.io || (socketModule.getIO && socketModule.getIO());
        if (io) {
            conversation.participants.forEach((pId) => {
                io.to(pId.toString()).emit("message:edited", {
                    _id: message._id.toString(),
                    conversationId: message.conversationId.toString(),
                    text: message.text,
                    isEdited: true,
                    updatedAt: message.updatedAt
                });
            });
        }
    }

    return res.status(200).json(new apiResponse(200, "Message edited successfully", message));
});

// Delete a message globally for all participants
export const deleteMessageForEveryone = asyncHandler(async (req, res, next) => {
    const { messageId } = req.params;
    const userId = req.user?._id;

    if (!userId || !mongoose.Types.ObjectId.isValid(userId)) {
        throw new apiError(401, "Unauthorized");
    }

    if (!messageId || !mongoose.Types.ObjectId.isValid(messageId)) {
        throw new apiError(400, "Invalid message ID format.");
    }

    const message = await Message.findById(messageId);
    if (!message) {
        throw new apiError(404, "Message not found");
    }

    if (message.senderId.toString() !== userId.toString()) {
        throw new apiError(403, "You can only delete your own messages");
    }

    message.text = "";
    message.imageUrl = "";
    message.attachmentUrl = "";
    message.attachmentType = "";
    message.attachmentMime = "";
    message.deletedForEveryone = true;
    await message.save();

    const conversation = await Conversation.findById(message.conversationId);
    if (conversation) {
        const io = socketModule.io || (socketModule.getIO && socketModule.getIO());
        if (io) {
            conversation.participants.forEach((pId) => {
                io.to(pId.toString()).emit("message:deletedForEveryone", {
                    _id: message._id.toString(),
                    conversationId: message.conversationId.toString(),
                    deletedForEveryone: true,
                    updatedAt: message.updatedAt
                });
            });
        }
    }

    return res.status(200).json(new apiResponse(200, "Message deleted for everyone", message));
});

// Hide a message from the caller's view (deletedForMe)
export const deleteMessageForMe = asyncHandler(async (req, res, next) => {
    const { messageId } = req.params;
    const userId = req.user?._id;

    if (!userId || !mongoose.Types.ObjectId.isValid(userId)) {
        throw new apiError(401, "Unauthorized");
    }

    if (!messageId || !mongoose.Types.ObjectId.isValid(messageId)) {
        throw new apiError(400, "Invalid message ID format.");
    }

    const message = await Message.findById(messageId);
    if (!message) {
        throw new apiError(404, "Message not found");
    }

    const conversation = await Conversation.findById(message.conversationId);
    if (!conversation) {
        throw new apiError(404, "Conversation not found");
    }

    const isParticipant = conversation.participants.some(p => p.toString() === userId.toString());
    if (!isParticipant) {
        throw new apiError(403, "You are not authorized to access this message");
    }

    if (!message.deletedForMe.includes(userId)) {
        message.deletedForMe.push(userId);
        await message.save();
    }

    const io = socketModule.io || (socketModule.getIO && socketModule.getIO());
    if (io) {
        io.to(userId.toString()).emit("message:deletedForMe", {
            _id: message._id.toString(),
            conversationId: message.conversationId.toString(),
        });
    }

    return res.status(200).json(new apiResponse(200, "Message deleted for you", message));
});

File: fastapi-backend/app/controllers/message_controller.py
import asyncio
import io
import logging
from datetime import datetime, timezone

import cloudinary.uploader
from firebase_admin import messaging

from app.config.firebase import get_firebase_app
from app.config.supabase import get_supabase
from app.constants.constants import MESSAGE_EDIT_WINDOW_MS
from app.services.ai_service import generate_ai_response
from app.socket.socket_manager import get_io
from app.utils.api_error import ApiError
from app.utils.api_response import ApiResponse
from app.utils.serializers import is_valid_uuid, serialize_message

logger = logging.getLogger(__name__)


async def _send_silent_push(token: str | None, sender_name: str, ciphertext: str, conversation_id: str) -> None:
    if not token:
        return
    try:
        get_firebase_app()
        await asyncio.to_thread(
            messaging.send,
            messaging.Message(
                data={
                    "senderName": sender_name,
                    "ciphertext": ciphertext or "",
                    "conversationId": conversation_id or "",
                },
                token=token,
            )
        )
    except Exception as exc:
        logger.error("Failed to send silent push notification: %s", exc)


async def get_messages(user: dict, conversation_id: str) -> ApiResponse:
    user_id = user.get("id")
    if not user_id or not is_valid_uuid(user_id):
        raise ApiError(401, "Unauthorized")
    if not conversation_id or not is_valid_uuid(conversation_id):
        raise ApiError(400, "Invalid conversation ID format.")

    supabase = get_supabase()
    conversation_result = supabase.table("conversations").select("*").eq("id", conversation_id).maybe_single().execute()
    conversation = getattr(conversation_result, "data", None) if conversation_result else None
    if not conversation:
        raise ApiError(404, "Conversation not found")

    if user_id not in (conversation.get("participants") or []):
        raise ApiError(403, "You are not authorized to view messages in this conversation")

    messages_result = (
        supabase.table("messages")
        .select("*")
        .eq("conversation_id", conversation_id)
        .order("created_at", desc=True)
        .limit(50)
        .execute()
    )

    filtered = [
        row
        for row in messages_result.data or []
        if user_id not in (row.get("deleted_for_me") or [])
    ]
    chronological = list(reversed(filtered))
    serialized = [serialize_message(row) for row in chronological]
    return ApiResponse(200, "Messages fetched successfully", serialized)


async def clear_messages(user: dict) -> ApiResponse:
    user_id = user.get("id")
    if not user_id or not is_valid_uuid(user_id):
        raise ApiError(401, "Unauthorized")

    supabase = get_supabase()
    conversations_result = supabase.table("conversations").select("id").contains("participants", [user_id]).execute()
    conversation_ids = [row["id"] for row in conversations_result.data or []]

    if not conversation_ids:
        return ApiResponse(200, "No messages to clear.", {"deletedCount": 0})

    deleted_count = 0
    for conversation_id in conversation_ids:
        messages_result = supabase.table("messages").select("id").eq("conversation_id", conversation_id).execute()
        message_ids = [row["id"] for row in messages_result.data or []]
        if message_ids:
            supabase.table("messages").delete().in_("id", message_ids).execute()
            deleted_count += len(message_ids)

    for conversation_id in conversation_ids:
        supabase.table("conversations").update({"last_message_id": None}).eq("id", conversation_id).execute()

    return ApiResponse(200, "Chat history cleared successfully.", {"deletedCount": deleted_count})


async def upload_attachment(file_bytes: bytes) -> ApiResponse:
    if not file_bytes:
        raise ApiError(400, "No file provided.")

    upload_response = await asyncio.to_thread(
        cloudinary.uploader.upload,
        io.BytesIO(file_bytes),
        resource_type="raw",
        folder="zync_secure_media",
    )
    return ApiResponse(200, "File uploaded successfully", {"url": upload_response["secure_url"]})


async def send_message(user: dict, conversation_id: str, body) -> ApiResponse:
    sender_id = user.get("id")
    if not sender_id or not is_valid_uuid(sender_id):
        raise ApiError(401, "Unauthorized")
    if not conversation_id or not is_valid_uuid(conversation_id):
        raise ApiError(400, "Conversation ID is missing or invalid.")

    has_text = isinstance(body.text, str) and body.text.strip()
    if not has_text and not body.image and not body.attachment_url:
        raise ApiError(400, "Message must contain text, an image, or an attachment.")

    supabase = get_supabase()
    conversation_result = supabase.table("conversations").select("*").eq("id", conversation_id).maybe_single().execute()
    conversation = getattr(conversation_result, "data", None) if conversation_result else None
    if not conversation:
        raise ApiError(404, "Conversation not found")
    if sender_id not in (conversation.get("participants") or []):
        raise ApiError(403, "You are not authorized to send messages")

    image_url = ""
    if body.image:
        upload_response = await asyncio.to_thread(cloudinary.uploader.upload, body.image, folder="zync_messages")
        image_url = upload_response["secure_url"]

    message_result = (
        supabase.table("messages")
        .insert(
            {
                "conversation_id": conversation_id,
                "sender_id": sender_id,
                "text": body.text.strip() if body.text else "",
                "image_url": image_url,
                "attachment_url": body.attachment_url or "",
                "attachment_type": body.attachment_type or "",
                "attachment_mime": body.attachment_mime or "",
            }
        )
        .execute()
    )
    new_message_row = (message_result.data or [{}])[0]
    now = datetime.now(timezone.utc).isoformat()
    supabase.table("conversations").update(
        {"last_message_at": now, "last_message_id": new_message_row["id"]}
    ).eq("id", conversation_id).execute()

    other_participant_ids = [pid for pid in conversation.get("participants") or [] if pid != sender_id]
    receivers_result = supabase.table("users").select("*").in_("id", other_participant_ids).execute()
    receivers = receivers_result.data or []
    io = get_io()
    new_message = serialize_message(new_message_row)

    for receiver in receivers:
        if receiver.get("is_ai"):
            plain_text_prompt = body.text or ""
            if body.image:
                plain_text_prompt += "\n[User also attached an image payload]"

            ai_response_text = await generate_ai_response(plain_text_prompt)
            ai_message_result = (
                supabase.table("messages")
                .insert(
                    {
                        "sender_id": receiver["id"],
                        "conversation_id": conversation_id,
                        "text": ai_response_text,
                        "image_url": "",
                    }
                )
                .execute()
            )
            ai_message_row = (ai_message_result.data or [{}])[0]
            supabase.table("conversations").update(
                {"last_message_at": datetime.now(timezone.utc).isoformat(), "last_message_id": ai_message_row["id"]}
            ).eq("id", conversation_id).execute()

            ai_payload = serialize_message(ai_message_row)
            if io:
                await io.emit("newMessage", ai_payload, room=sender_id)

            if user.get("fcmToken"):
                await _send_silent_push(
                    user["fcmToken"],
                    receiver.get("display_name") or receiver.get("username") or "AI Assistant",
                    ai_message_row.get("text", ""),
                    conversation_id,
                )
        else:
            if io:
                await io.emit("newMessage", new_message, room=receiver["id"])
            if receiver.get("fcm_token"):
                await _send_silent_push(
                    receiver["fcm_token"],
                    user.get("displayName") or user.get("username") or "Someone",
                    new_message_row.get("text", ""),
                    conversation_id,
                )

    return ApiResponse(201, "Message sent successfully", new_message)


async def edit_message(user: dict, message_id: str, text: str) -> ApiResponse:
    user_id = user.get("id")
    if not user_id or not is_valid_uuid(user_id):
        raise ApiError(401, "Unauthorized")
    if not message_id or not is_valid_uuid(message_id):
        raise ApiError(400, "Invalid message ID format.")

    supabase = get_supabase()
    message_result = supabase.table("messages").select("*").eq("id", message_id).maybe_single().execute()
    message = getattr(message_result, "data", None) if message_result else None
    if not message:
        raise ApiError(404, "Message not found")
    if message["sender_id"] != user_id:
        raise ApiError(403, "You can only edit your own messages")

    created_at = datetime.fromisoformat(message["created_at"].replace("Z", "+00:00"))
    if (datetime.now(timezone.utc) - created_at).total_seconds() * 1000 >= MESSAGE_EDIT_WINDOW_MS:
        raise ApiError(403, "Message cannot be edited after the edit window has elapsed")

    update_result = (
        supabase.table("messages")
        .update({"text": text.strip(), "is_edited": True})
        .eq("id", message_id)
        .execute()
    )
    updated = serialize_message((update_result.data or [{}])[0])

    conversation_result = (
        supabase.table("conversations").select("participants").eq("id", message["conversation_id"]).maybe_single().execute()
    )
    conversation = getattr(conversation_result, "data", None) if conversation_result else None
    io = get_io()
    if conversation and io:
        for participant_id in conversation.get("participants") or []:
            await io.emit(
                "message:edited",
                {
                    "_id": updated["id"],
                    "conversationId": updated["conversationId"],
                    "text": updated["text"],
                    "isEdited": True,
                    "updatedAt": updated["updatedAt"],
                },
                room=participant_id,
            )

    return ApiResponse(200, "Message edited successfully", updated)


async def delete_message_for_everyone(user: dict, message_id: str) -> ApiResponse:
    user_id = user.get("id")
    if not user_id or not is_valid_uuid(user_id):
        raise ApiError(401, "Unauthorized")
    if not message_id or not is_valid_uuid(message_id):
        raise ApiError(400, "Invalid message ID format.")

    supabase = get_supabase()
    message_result = supabase.table("messages").select("*").eq("id", message_id).maybe_single().execute()
    message = getattr(message_result, "data", None) if message_result else None
    if not message:
        raise ApiError(404, "Message not found")
    if message["sender_id"] != user_id:
        raise ApiError(403, "You can only delete your own messages")

    update_result = (
        supabase.table("messages")
        .update(
            {
                "text": "",
                "image_url": "",
                "attachment_url": "",
                "attachment_type": "",
                "attachment_mime": "",
                "deleted_for_everyone": True,
            }
        )
        .eq("id", message_id)
        .execute()
    )
    updated = serialize_message((update_result.data or [{}])[0])

    conversation_result = (
        supabase.table("conversations").select("participants").eq("id", message["conversation_id"]).maybe_single().execute()
    )
    conversation = getattr(conversation_result, "data", None) if conversation_result else None
    io = get_io()
    if conversation and io:
        for participant_id in conversation.get("participants") or []:
            await io.emit(
                "message:deletedForEveryone",
                {
                    "_id": updated["id"],
                    "conversationId": updated["conversationId"],
                    "deletedForEveryone": True,
                    "updatedAt": updated["updatedAt"],
                },
                room=participant_id,
            )

    return ApiResponse(200, "Message deleted for everyone", updated)


async def delete_message_for_me(user: dict, message_id: str) -> ApiResponse:
    user_id = user.get("id")
    if not user_id or not is_valid_uuid(user_id):
        raise ApiError(401, "Unauthorized")
    if not message_id or not is_valid_uuid(message_id):
        raise ApiError(400, "Invalid message ID format.")

    supabase = get_supabase()
    message_result = supabase.table("messages").select("*").eq("id", message_id).maybe_single().execute()
    message = getattr(message_result, "data", None) if message_result else None
    if not message:
        raise ApiError(404, "Message not found")

    conversation_result = (
        supabase.table("conversations").select("*").eq("id", message["conversation_id"]).maybe_single().execute()
    )
    conversation = getattr(conversation_result, "data", None) if conversation_result else None
    if not conversation:
        raise ApiError(404, "Conversation not found")
    if user_id not in (conversation.get("participants") or []):
        raise ApiError(403, "You are not authorized to access this message")

    deleted_for_me = message.get("deleted_for_me") or []
    if user_id not in deleted_for_me:
        deleted_for_me.append(user_id)
        supabase.table("messages").update({"deleted_for_me": deleted_for_me}).eq("id", message_id).execute()
        message["deleted_for_me"] = deleted_for_me

    io = get_io()
    if io:
        await io.emit(
            "message:deletedForMe",
            {"_id": message_id, "conversationId": message["conversation_id"]},
            room=user_id,
        )

    return ApiResponse(200, "Message deleted for you", serialize_message(message))
